test/test_integrate.py

```diff
@@ -398,6 +398,30 @@ def g(t, y, args):
             assert tree_allclose(sol1.derivative(ti), -sol2.derivative(-ti))
 
 
+def test_reverse_time_save():
```

Comment:
As above I think it's worth including other kinds of `saveat` here, which include `t0`, `t1` etc, to be sure that these cases don't slip through.

Reply:
Expanded to include all the flags of saveat. Note that I specify max steps because if that is greater and steps is being logged, it multiples the jnp.inf by direction at the end which results in a mismatch. I don't see a great way to get the custom function to match that and given that those infinities are just padding/holding space, I don't consider it an actual issue

Reply:
Hmm, that surprises me. The `jnp.inf` should just be the initial buffer values, I don't think they're touched afterwards. In particular I don't think the choice of `direction` should affect them?

Other than that this PR LGTM!

Reply:
Sepcifically, the output ys vs ts are [3.9 3.8 3.7 3.6 3.5 3.4 3.3 3.2 3.1 3.  2.9 2.8 2.7 2.6 2.5 2.4 2.3 2.2
 2.1 2.  1.9 1.8 1.7 1.6 1.5 1.4 1.3 1.2 1.1 1.  0.9 0.8 0.7 0.6 0.5 0.4
 0.3 inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf
 inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf
 inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf inf
 inf inf inf inf inf inf inf inf inf inf]

vs

 [ 3.9  3.8  3.7  3.6  3.5  3.4  3.3  3.2  3.1  3.   2.9  2.8  2.7  2.6
  2.5  2.4  2.3  2.2  2.1  2.   1.9  1.8  1.7  1.6  1.5  1.4  1.3  1.2
  1.1  1.   0.9  0.8  0.7  0.6  0.5  0.4  0.3 -inf -inf -inf -inf -inf
 -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf
 -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf
 -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf
 -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf
 -inf -inf]

The ts get a direction multiplied by them at the end, but it is initialized to `inf` without getting direction multiplied. I think this might be a bug, so I fixed it as well and now all the ts inf get initialized with respect to the direction (        ts = jnp.full(out_size, direction * jnp.inf, dtype=time_dtype)). Maybe not a bug since these `inf` are technically meaningless, but at least now the tests pass no problem and I can remove max steps.

Reply:
Hmm, I guess that actually breaks `assert tree_allclose(sol1.ts, -cast(Array, sol2.ts), equal_nan=True)` for steps=True because then the infs are not opposite signs. Not sure if they should be or not, since they are basically superfluous. I will defer to whatever you think. I can revert the last commit and keep max steps and the infs will be sign flipped, or I can make a change to what this failing test expects and the infs will always be positive.

Reply:
Right. So I think this is indicative of a totally separate bug: in some cases we've been returning `-inf` all along:
```python
diffrax.diffeqsolve(diffrax.ODETerm(lambda t, y, args: -y), diffrax.Tsit5(), 1, 0, -0.1, 1, saveat=diffrax.SaveAt(steps=True), max_steps=50).ts
```
This shouldn't be the case. I think we should always return `+inf` for our padding!

It looks like the culprit is as you've identified, namely that we flip things here:

https://github.com/patrick-kidger/diffrax/blob/c6cc85c92b0e14e24092830584d0397a1990d091/diffrax/_integrate.py#L1018-L1020

I think we should adjust the failing test you've highlighted, so that the infs are always positive.

Reply:
Great, I changed that test, and all padding should be positive now.